lib/redux/actions/async.js

```diff
@@ -0,0 +1,560 @@
+        }
+        // remove env-switching context menu after login
+        if (typeof chrome !== 'undefined') {
+          services.log('Removing Chrome context menu');
+          chrome.contextMenus.removeAll();
+        }
+        dispatch(syncActions.initSuccess());
+        dispatch(doVersionCheck());
+        dispatch(syncActions.setUserInfoFromToken({
+          user: results[0],
+          profile: results[1],
+          memberships: results[2]
+        }));
+        const { users } = getState();
+        if (users.uploadTargetUser !== null) {
+          dispatch(syncActions.setBlipViewDataUrl(
+            api.makeBlipUrl(viewDataPathForUser(users.uploadTargetUser))
+          ));
+        }
+        dispatch(retrieveTargetsFromStorage());
+      });
+    });
+  };
+}
+
+function viewDataPathForUser(uploadTargetUser) {
+  return `/patients/${uploadTargetUser}/data`;
+}
+
+export function doLogin(creds, opts) {
+  return (dispatch, getState) => {
+    const { api } = services;
+    dispatch(syncActions.loginRequest());
+
+    async.series([
+      api.user.login.bind(null, creds, opts),
+      api.user.profile,
+      api.user.getUploadGroups
+    ], (err, results) => {
+      if (err) {
+        return dispatch(syncActions.loginFailure(err.status));
+      }
+      // remove env-switching context menu after login
+      if (typeof chrome !== 'undefined') {
+        services.log('Removing Chrome context menu');
+        chrome.contextMenus.removeAll();
+      }
+      dispatch(syncActions.loginSuccess({
+        user: results[0].user,
+        profile: results[1],
+        memberships: results[2]
+      }));
+      const { users } = getState();
+      if (users.uploadTargetUser !== null) {
+        dispatch(syncActions.setBlipViewDataUrl(
+          api.makeBlipUrl(viewDataPathForUser(users.uploadTargetUser))
+        ));
+      }
+      dispatch(retrieveTargetsFromStorage());
+    });
+  };
+}
+
+export function doLogout() {
+  return (dispatch) => {
+    const { api } = services;
+    dispatch(syncActions.logoutRequest());
+    api.user.logout((err) => {
+      if (err) {
+        dispatch(syncActions.logoutFailure());
+        dispatch(syncActions.setPage(pages.LOGIN, actionSources.USER));
+      }
+      else {
+        dispatch(syncActions.logoutSuccess());
+        dispatch(syncActions.setPage(pages.LOGIN, actionSources.USER));
+      }
+    });
+  };
+}
+
+function getUtc(utc) {
+  return _.isEmpty(utc) ? sundial.utcDateString() : utc;
+}
+
+function makeProgressFn(dispatch, step, percentage) {
+  return (step, percentage) => {
+    dispatch(syncActions.uploadProgress(step, percentage));
+  };
+}
+
+function makeUploadCb(dispatch, getState, errCode, utc) {
+  return (err, recs) => {
+    const { devices, uploads, users, version } = getState();
+    const targetDevice = devices[uploads.uploadInProgress.pathToUpload[1]];
+    if (err) {
+      const serverDown = 'Origin is not allowed by Access-Control-Allow-Origin';
```

Comment:
From the error message, and its relation to CORS, this seems to me like a funny variable name. Something like serverError, might be better?